main/src/library/Graph.flix

```diff
@@ -50,4 +50,47 @@ namespace Graph {
         List.toSet(x)
     }
 
+    ///
+    /// Returns the strongly connected components of the directed graph `g`.
+    ///
+    pub def stronglyConnectedComponents(g: m[(t, t)]): List[Set[t]] with Foldable[m], Boxable[t] =
+        let edges = inject g into Edge;
+        let nodes = #{
+            Node(x) :- Edge(x, _).
+            Node(x) :- Edge(_, x).
+        };
+        let reachability = #{
+            // All nodes can reach themselves.
+            Reachable(n1, n1) :- Edge(n1, _).
+            Reachable(n2, n2) :- Edge(_, n2).
+            // If `n1` can reach `m` and there is an edge from `m` to `n2`
+            // then `n1` can also reach `n2`. This adds all node pairs to
+            // the relational that are reachable using any number of nodes.
+            Reachable(n1, n2) :- Reachable(n1, m), Edge(m, n2).
+        };
+        let connected = #{
+            // If `n1` can reach `n2` and `n2` can reach `n1`
+            // then they are part of the same strongly connected component.
+            Connected(n1, n2) :- Reachable(n1, n2), Reachable(n2, n1).
+        };
+
+        let allNodes = query edges, nodes select x from Node(x);
+
+        // Find every node that is part of the strongly connected component with `x`.
+        let f = (x) -> {
+            query edges, reachability, connected select dst from Connected(src, dst) where src == x
+        };
+
+        let f1 = (acc, x) -> {
```

Comment:
I'm less certain on how to do this part in Datalog. Maybe we could use the same trick just on the sets
```
Scc(;Set#{scc}) :- Node(a), Connected(a; scc)
```
Now the facts of `Scc` will have a single fact which is the set of set of nodes that correspond to the sccs, Maybe `Set[Set[t]]` is also suited for the return type since the components are not inherently ordered.

I'm not sure how slow/fast this is

Reply:
I'll try it out and maybe read up on the lattice stuff on monday